crates/bevy_macro_utils/src/lib.rs

```diff
@@ -118,12 +118,41 @@ pub fn derive_label(input: syn::DeriveInput, trait_path: &syn::Path) -> TokenStr
         where_token: Default::default(),
         predicates: Default::default(),
     });
-    where_clause.predicates.push(syn::parse2(quote! { Self: Eq + ::std::fmt::Debug + ::std::hash::Hash + Clone + Send + Sync + 'static }).unwrap());
+    where_clause
+        .predicates
+        .push(syn::parse2(quote! { Self: 'static }).unwrap());
+
+    let as_str = match input.data {
+        syn::Data::Struct(d) => match d.fields {
+            syn::Fields::Unit => {
+                let lit = ident.to_string();
```

Comment:
Given that `as_str()` results are used for comparison, it seems like we could cut down on Eq + Hash costs by:
1. Using an empty string for Unit structs
2. Only using the enum variant name for Enums

For debugging purposes, the full type name could be used. It would mean storing a separate type_name string, which increases the size of each label by an extra `&'static str`. But its worth checking to see if its worth it.

Reply:
That's definitely something to look into but I think it would make sense to discuss it as a separate PR due to all the tradeoffs you have to make with micro-optimizations like that.

Reply:
I thought of a simple, zero-cost optimization for this. When one side of an `==` operation is a concrete `impl SystemLabel` and the other is a type erased `SystemLabelId`, we can allow the concrete type to perform a custom comparison. For unit types, this means really fast comparisons by ignoring the string. For enums, this would mean chopping off the type name part of the string and only comparing the variant name.

Granted this wouldn't be too meaningful, since most label comparisons are done between two type erased labels. Although there are some, such as when inserting a new `Stage`, I believe.

Reply:
I did end up trying this change, and the benchmarks were not promising. Best improvement was -2%, worst regression was +5%. The others weren't affected at all.